quinn-h3/src/body.rs

```diff
@@ -159,28 +164,66 @@ impl Future for RecvBody {
                     }
                     _ => return Err(Error::peer("invalid frame type in data")),
                 },
-                RecvBodyState::Decoding(ref frame) => {
-                    let result = {
-                        let conn = &mut self.conn.h3.lock().unwrap().inner;
-                        conn.decode_header(self.stream_id, frame)
-                    };
-
-                    match result {
-                        Ok(None) => return Ok(Async::NotReady),
-                        Err(e) => {
-                            return Err(Error::peer(format!("decoding header failed: {:?}", e)))
-                        }
-                        Ok(Some(decoded)) => {
-                            self.state = RecvBodyState::Finished;
-                            return Ok(Async::Ready((
-                                try_take(&mut self.body, "body absent")?,
-                                Some(decoded.into_fields()),
-                            )));
-                        }
-                    }
+                RecvBodyState::Decoding(ref mut trailer) => {
+                    let trailer = try_ready!(trailer.poll());
+                    self.state = RecvBodyState::Finished;
+                    return Ok(Async::Ready((
+                        try_take(&mut self.body, "body absent")?,
+                        Some(trailer.into_fields()),
+                    )));
                 }
                 _ => return Err(Error::Poll),
             }
         }
     }
 }
+
+pub struct RecvBodyStream {
+    recv: FrameStream,
+    trailers: Option<HeadersFrame>,
+    conn: ConnectionRef,
+    stream_id: StreamId,
+}
+
+impl RecvBodyStream {
+    pub(crate) fn new(recv: FrameStream, conn: ConnectionRef, stream_id: StreamId) -> Self {
+        RecvBodyStream {
+            recv,
+            conn,
+            stream_id,
+            trailers: None,
+        }
+    }
+
+    pub fn has_trailers(&self) -> bool {
+        self.trailers.is_some()
+    }
+
+    pub fn trailers(self) -> Option<DecodeHeaders> {
+        match self.trailers {
+            None => None,
+            Some(t) => Some(DecodeHeaders::new(t, self.conn, self.stream_id)),
+        }
+    }
+}
+
+impl Stream for RecvBodyStream {
+    type Item = Bytes;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match try_ready!(self.recv.poll()) {
+            Some(HttpFrame::Data(d)) => {
+                Ok(Async::Ready(Some(d.payload)))
+            }
+            Some(HttpFrame::Headers(d)) => {
+                self.trailers = Some(d);
+                Ok(Async::Ready(None))
```

Comment:
If `Headers` can ever be received prior to additional `Data`, this could cause a hung future. I imagine that's forbidden by the h3 spec, but what happens if a peer is ill-behaved?

Reply:
From what I understand, the Stream will only appear empty to the user. The additional data will be dropped along with `self.recv`.

I'm not sure how to make polling again for another `DataFrame` after one `HeadersFrame` end up with a correct behavior.

Reply:
You're right, I was confused; this correctly terminates the stream, and assuming the sender handles a dropped receiver gracefully all is well.